BUG allow subclasses in MaskedArray ufuncs -- for non-ndarray _data

mhvk opened the pull request:
This is a follow-on to #3900, making two furthe changes to tryto reach the ultimate goal of `MaskedArray` behaving sensibly for subclasses. 
1. Ensure that for two-argument ufunc's, the data in the subclass are no longer converted to `ndarray`; rather, a possible subclass is kept. This is particularly relevant for my work on a `MaskedQuantity` class, where `_data` contains a `Quantity` which has not just a value (the ndarray) but also a unit.
2. Avoid the creation of `object` arrays, so that non-array Classes with an `__array_priority__` set and with a reverse method defined properly lead to `NotImplemented` (i.e., making the behaviour match that of regular arrays).

In order for this to work, I
- removed `subok=False` in the `getdata` calls in the two `BinaryOperation` classes
- slightly changed the method used to keep previous values when they are masked, including allowing this to fail (e.g., when I multiply two masked quantities in meters, the result is an area, and hence it is senseless to keep the original values, and quantities raise an appropriate UnitError). (Without this, the matrix subclass tests would not pass).
- slightly changed the order, with result masks only calculated if the primary calculation succeeded.
- slightly changed `getdata` to avoid the creation of `object` arrays, and `getmaskedarray` to deal with the consequences.

I also added a test case, which perhaps best makes clear why this is useful. 

CC @charris -- I consider the above a correction of a bug, but realise this is subjective; it could also be ENH, even though the API does not change. It would be wonderful if this could still make it to 1.8.

mhvk committed:
BUG Allow subclasses in MaskedArray ufuncs -- for non-ndarray _data

charris commented:
@mhvk Getting a lot of these
```
======================================================================
ERROR: test_addsumprod (test_core.TestMaskedArrayArithmetic)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/virtualenv/python2.7.6/lib/python2.7/site-packages/numpy/ma/tests/test_core.py", line 931, in test_addsumprod
    assert_equal(np.add.reduce(x), add.reduce(x))
  File "/home/travis/virtualenv/python2.7.6/lib/python2.7/site-packages/numpy/ma/core.py", line 989, in reduce
    masked_tr._update_from(tr)
AttributeError: 'numpy.float64' object has no attribute '_update_from'
```

mhvk commented:
OK, will try to look at this in more detail soon -- this is a while back...

charris commented:
Looking at this more for 1.10.0. The problem looks to be an array priority thing and thus overlaps some of the other masked array PRs. Perhaps the `_MaskedBinaryOperation` should be modified to deal with that? I don't like the idea of changing `getdata`, but it seems to be prematurely called, or maybe should not be used at all, in  `_MaskedBinaryOperation`.

Anyway, I'd like to work something helpful out for 1.10.0. There a couple of other PRs from AstroPy and I'd like to deal with all of them, but need some help.

mhvk commented:
@charris - thanks for pointing me to #5227. I tried the trial implementation of taking account of `__array_priority__` of @abalkin in #5230 and that solves the problem where `<masked-array> * <unit>` made a masked array of object dtype. With that, I think that the somewhat problematic edits to `getdata` would no longer be needed.

However, #5230 does not address that subclasses should be passed on (the main goal of this PR). Since it is simpler, though, maybe it should be done first?

charris commented:
@mhvk Where do you think this stands with respect to the `__numpy_ufunc__` discussion? Also, needs a fix if this is going to be considered for 1.10.

mhvk commented:
@charris - This is in addition to the discussion on `__array_priority__` -- ideally, I would rebase after #5864 is merged, and focus just on the subclass issue (i.e., remove the changes to `getdata`).

charris commented:
@mhvk Thanks for the update. I'll let this wait on #5864 then.

charris commented on numpy/ma/core.py:
```diff
@@ -3692,7 +3709,7 @@ def __add__(self, other):
     def __radd__(self, other):
         "Add other to self, and return a new masked array."
-        return add(self, other)
+        return add(other, self)
```
So addition may not be commutative. That is also the case with `__numpy_ufunc__`. Could you put an explanatory comment here so that some well meaning individual doesn't "fix" it in the future.

charris commented on numpy/ma/core.py:
```diff
@@ -3708,7 +3725,7 @@ def __mul__(self, other):
     def __rmul__(self, other):
         "Multiply other by self, and return a new masked array."
-        return multiply(self, other)
+        return multiply(other, self)
```
Also a comment here.

charris commented:
Could use a couple of comments.

charris commented:
@mhvk #5864 is in.

mhvk commented:
@charris - thanks, I'll try to look at this and possible tests for #4617 as soon as I can.

mhvk commented:
@charris - OK, I rebased this one, which meant that, as hoped, the changes to `getdata` could be reverted. Furthermore, with #5864 in, there no longer is a need to pass down `NotImplemented` from within the ufunc implementations (more in the spirit of what we seem to be getting to for #5844, so good in that sense). So, all that's left is now a rewrite that ensures that subclasses get to tell how the ufuncs are done. Which is quite clean and nicely stand-alone.

Apart from that, I added the comments to `__radd__` and `__rmul__` you requested and one test for #4617 (can do more if you think it is important but I ran out of time today).

charris commented:
Hmm, this test failure looks potentially real (but maybe not repeatable). I'm going to restart the test but keeping this for reference. The `numpy/ma/testutils.py` is the only thing that looks related.
```
FAIL: test_io.test_load_refcount

----------------------------------------------------------------------

Traceback (most recent call last):

  File "/home/travis/build/numpy/numpy/builds/venv/lib/python2.7/site-packages/nose/case.py", line 197, in runTest

    self.test(*self.arg)

  File "/home/travis/build/numpy/numpy/builds/venv/lib/python2.7/site-packages/numpy/lib/tests/test_io.py", line 1920, in test_load_refcount

    assert_equal(n_before, n_after)

  File "/home/travis/build/numpy/numpy/builds/venv/lib/python2.7/site-packages/numpy/ma/testutils.py", line 100, in assert_equal

    raise AssertionError(msg)

AssertionError: 

Items are not equal:

 ACTUAL: 55930

 DESIRED: 55911
```

mhvk commented:
@charris - yes, I had seen that very weird test result, and was going to ask to restart, as it seemed unlikely any change in the python code would lead to a refcount problem. But you beat me to it... oddly, it seems irreproduceable (though at least it means it was not due to this PR).

charris commented:
Well, lets get this in. Thanks @mhvk.

argriffing commented:
A possibly related issue on the mailing list: https://mail.scipy.org/pipermail/numpy-discussion/2016-February/074940.html
